### PR TITLE
GDKotlin Rework: Jvm loading

### PR DESCRIPTION
**Pull request by CedNaru**
Rework of the whole process of loading the JVM, the bootstrap and then the usercode.

Made some dirty fix to Bootstrap.kt, which will quickly be replaced anyway in the future CPP reloading rework, for now it's there just to run.
Right now, I make the whole engine crash if something goes wrong in the initialization process. I'll change that to proper reporting and fallback to a Kotlin-less Godot in the next PR.

### CODE REVIEW

**Comment by CedNaru** [comment]
Tested on Windows with the following configurations:
- Editor with Embedded JVM
- Editor with JAVA_HOME only
- Export with JVM
- Export with Graal native image

**Comment by CedNaru** [comment]
Tested the Export on Android.

**Review by chippmann — CHANGES_REQUESTED** [review]
Tested on linux:
- Editor with `JAVA_HOME` -> also with reloading
- Editor with embedded jvm -> also with reloading
- Debug export templates embedded jvm
- Release export templates embedded jvm
- Debug export templates graaal vm native image
- Release export templates graaal vm native image

Apart from the encountered issues regarding configuration file parsing and exporting mentioned on discord, everything seems to work well.

Awesome job!

Will continue testing throughout the day on a pet project

**Review by chippmann — APPROVED** [review]

File: src/lifecycle/class_loader.cpp [hunk 1]
```diff
@@ -2,70 +2,82 @@
...
     return url;
 }
+#endif
 
-jni::JObject ClassLoader::provide_loader(jni::Env& env, const String& full_jar_path, const jni::JObject& p_parent_loader) {
+ClassLoader* ClassLoader::create_instance(jni::Env& env, const String& full_jar_path, const jni::JObject& p_parent_loader) {
 #ifdef __ANDROID__
-    jni::JClass class_loader_cls {env.find_class("dalvik/system/DexClassLoader")};
-    jni::MethodId ctor {class_loader_cls.get_constructor_method_id(env, "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/ClassLoader;)V")};
+    jni ::JClass class_loader_cls {env.find_class("dalvik/system/DexClassLoader")};
```
Comment (piiertho) [inline src/lifecycle/class_loader.cpp:34]: typo: `jni ::JClass` => `jni::JClass`

File: src/lifecycle/paths.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+static constexpr const char* BOOTSTRAP_FILE {"godot-bootstrap-dex.jar"};
+static constexpr const char* USER_CODE_FILE {"main-dex.jar"};
+static constexpr const char* GRAAL_NATIVE_IMAGE_FILE {""};
+
+#elif IOS_ENABLED
+
+static constexpr const char* RELATIVE_JVM_LIB_PATH {""};
+static constexpr const char* BOOTSTRAP_FILE {""};
+static constexpr const char* USER_CODE_FILE {""};
+static constexpr const char* GRAAL_NATIVE_IMAGE_FILE {"usercode.so"};
```
Comment (piiertho) [inline src/lifecycle/paths.h:42]: This is weird to define a path to a shared library on iOS

File: src/lifecycle/jvm_options.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#include "jvm_options.h"
+
+void JvmOptions::add_jni_checks() {
+    options.push_back("-Xcheck:jni");
+}
+
+void JvmOptions::add_debug_options(uint16_t p_port, String& p_address, bool p_wait) {
+    String jvm_debug_port = String::num_int64(p_port);
+
+    String suspend;
```
[the remark this answers is not in the document]
Comment (CedNaru) [reply src/lifecycle/jvm_options.cpp:11]: Good catch, I just copied that part from the previous code, didn't think of changing it.

**Comment by chippmann** [comment]
Tested on macos x86_64:
With graalvm native image:
- editor
- export release
- export debug

**Review by piiertho — APPROVED** [review]

**Pull request merged**